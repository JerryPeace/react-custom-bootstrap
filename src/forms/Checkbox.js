import PropTypes from 'prop-types';
import React from 'react';
import {Basic} from './Basic';
import classNames from 'classnames';
import FormGroup from './FormGroup';
import FormControls from './FormControls';
import LayoutHelper from './LayoutHelper';

class Checkbox extends Basic {

    static propTypes = {
        name: PropTypes.string,
        labelText: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
        inline: PropTypes.bool,
        labelAlign: PropTypes.string,
        standalone: PropTypes.bool,
        inputType: PropTypes.oneOf(['checkbox'])
    };

    static defaultProps = {
        bsClass: 'checkbox',
        inline: false,
        labelAlign: 'right',
        standalone: true,
        inputType: 'checkbox',
        wrapperClassName: 'col-xs-offset-2 col-xs-10'
    };

    constructor(props) {
        super(props);
        this.originalVal = ::this._initOriginalVal();
    }

    state = {
        checked: this._initOriginalChecked()
    }

    _initOriginalVal() {
        const checked = this._initOriginalChecked();
        return checked ? (this.props.value || checked) : null;
    }

    _initOriginalChecked() {
        let checked;
        if (typeof this.props.checked !== 'undefined') {
            checked = this.props.checked;
        } else if (typeof this.props.defaultChecked !== 'undefined') {
            checked = this.props.defaultChecked;
        } else {
            checked = false;
        }

        return checked;
    }

    _onChange(e) {
        const {onChange} = this.props;
        if (typeof this.props.checked === 'undefined') {
            this.setState({
                checked: this.getChecked()
            });
        }
        onChange && onChange(e);
    }

    getValue() {
        const checked = this.getChecked();
        return checked ? (this._inputEl.getAttribute('value') || checked) : null;
    }

    getChecked() {
        return this._inputEl.checked;
    }

    renderInput() {
        const {className, inputType, labelText, inline, labelAlign, isChecked, standalone, bsClass, wrapperClassName, ...rest} = this.props;

        return (
            <input
                className={classNames(className)}
                ref={(el) => this._inputEl = el}
                {...rest}
                onChange={::this._onChange}
                type={inputType}
                key={'checkbox'} />
        );
    }

    render() {
        const checked = typeof this.props.checked !== 'undefined' ? this.props.checked : this.state.checked;
        return (
            <LayoutHelper
                FieldBase={this.renderInput()}
                {...this.props}
                checked={checked} />
        );
    }
}

export default Checkbox;
