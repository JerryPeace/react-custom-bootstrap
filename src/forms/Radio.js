import PropTypes from 'prop-types';
import React from 'react';
import {Basic} from './Basic';
import classNames from 'classnames';
import FormGroup from './FormGroup';
import FormControls from './FormControls';
import Label from './Label';
import LayoutHelper from './LayoutHelper';

class Radio extends Basic {

    static propTypes = {
        bsClass: PropTypes.string,
        checked: PropTypes.bool,
        className: PropTypes.string,
        defaultChecked: PropTypes.bool,
        inline: PropTypes.bool,
        inputType: PropTypes.oneOf(['radio']),
        labelAlign: PropTypes.string,
        labelText: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
        onChange: PropTypes.func,
        standalone: PropTypes.bool,
        value: PropTypes.any,
        wrapperClassName: PropTypes.string
    };

    static defaultProps = {
        bsClass: 'radio',
        inline: false,
        inputType: 'radio',
        labelAlign: 'right',
        standalone: true,
        wrapperClassName: 'col-xs-offset-2 col-xs-10'
    };

    constructor(props) {
        super(props);
        this.originalVal = ::this._initOriginalVal();
    }

    state ={
        checked: this._initOriginalChecked()
    }

    _initOriginalVal() {
        const checked = this._initOriginalChecked();
        return checked ? (this.props.value || checked) : null;
    }

    _initOriginalChecked() {
        let checked;
        if (typeof this.props.checked !== 'undefined') {
            checked = this.props.checked;
        } else if (typeof this.props.defaultChecked !== 'undefined') {
            checked = this.props.defaultChecked;
        } else {
            checked = false;
        }

        return checked;
    }

    _onChange(e) {
        const {onChange, value} = this.props;
        if (typeof this.props.checked === 'undefined') {
            this.setState({
                checked: this.getChecked()
            });
        }
        onChange && onChange(e, value);
    }

    getValue() {
        const checked = this.getChecked();
        return checked ? (this._inputEl.getAttribute('value') || checked) : null;
    }

    getChecked() {
        return this._inputEl.checked;
    }

    renderInput() {
        const {className, inputType, labelText, inline, labelAlign, standalone, bsClass, wrapperClassName, ...rest} = this.props;

        return (
            <input
                ref={(el) => this._inputEl = el}
                {...rest}
                onChange={::this._onChange}
                type={inputType}
                key={'radio'} />
        );
    }

    render() {
        const checked = typeof this.props.checked !== 'undefined' ? this.props.checked : this.state.checked;
        return (
            <LayoutHelper
                FieldBase={this.renderInput()}
                {...this.props}
                checked={checked} />
        );
    }
}

export default Radio;
