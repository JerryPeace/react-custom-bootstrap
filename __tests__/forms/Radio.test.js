import React from 'react';
import { mount } from 'enzyme';
import toJSON from'enzyme-to-json';
import Radio from 'forms/Radio';
import Checkbox from 'forms/Checkbox';

describe('Radio', () => {
    let mockChange;
    let radio;
    let checkbox;
    let ins;

    describe('with default props', () => {
        beforeEach(() => {
            mockChange = jest.fn();
            radio = mount(<Radio singleCheckboxRadio={true} onChange={mockChange} />);
        });

        test('should render corrently', () => {
            expect(toJSON(radio)).toMatchSnapshot();
        });

        test('click input would trigger callback', () => {
            radio.find('input').simulate('change');
            expect(mockChange.mock.calls.length).toBe(1);
        });
    });

    describe('with checked props', () => {
        beforeEach(() => {
            mockChange = jest.fn()
            radio = mount(<Radio singleCheckboxRadio={true} onChange={mockChange} checked={false} />);
        });

        test('should render corrently', () => {
            radio.update();
            expect(toJSON(radio)).toMatchSnapshot();
        });

        test('after checked changed should render correctly', () => {
            radio.setProps({ checked: true });
            radio.update();
            expect(toJSON(radio)).toMatchSnapshot();
        });
    });

});
