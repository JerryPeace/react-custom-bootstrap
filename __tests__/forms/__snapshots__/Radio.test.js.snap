// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio with checked props after checked changed should render correctly 1`] = `
<Radio
  bsClass="radio"
  checked={true}
  inline={false}
  inputType="radio"
  labelAlign="right"
  onChange={[Function]}
  singleCheckboxRadio={true}
  standalone={true}
  wrapperClassName="col-xs-offset-2 col-xs-10"
>
  <LayoutHelper
    FieldBase={
      <input
        checked={true}
        onChange={[Function]}
        singleCheckboxRadio={true}
        type="radio"
      />
    }
    bsClass="radio"
    checked={true}
    inline={false}
    inputType="radio"
    labelAlign="right"
    onChange={[Function]}
    singleCheckboxRadio={true}
    standalone={true}
    wrapperClassName="col-xs-offset-2 col-xs-10"
  >
    <div
      className="radio"
    >
      <Label
        FieldBase={
          <input
            checked={true}
            onChange={[Function]}
            singleCheckboxRadio={true}
            type="radio"
          />
        }
        className="checked checkbox-radio-label radio-wrapper"
        inline={false}
        inputType="radio"
        labelAlign="right"
        onChange={[Function]}
        wrapperClassName="col-xs-offset-2 col-xs-10"
      >
        <label
          className="checked checkbox-radio-label radio-wrapper"
        >
          <input
            checked={true}
            onChange={[Function]}
            singleCheckboxRadio={true}
            type="radio"
          />
        </label>
      </Label>
    </div>
  </LayoutHelper>
</Radio>
`;

exports[`Radio with checked props should render corrently 1`] = `
<Radio
  bsClass="radio"
  checked={false}
  inline={false}
  inputType="radio"
  labelAlign="right"
  onChange={[Function]}
  singleCheckboxRadio={true}
  standalone={true}
  wrapperClassName="col-xs-offset-2 col-xs-10"
>
  <LayoutHelper
    FieldBase={
      <input
        checked={false}
        onChange={[Function]}
        singleCheckboxRadio={true}
        type="radio"
      />
    }
    bsClass="radio"
    checked={false}
    inline={false}
    inputType="radio"
    labelAlign="right"
    onChange={[Function]}
    singleCheckboxRadio={true}
    standalone={true}
    wrapperClassName="col-xs-offset-2 col-xs-10"
  >
    <div
      className="radio"
    >
      <Label
        FieldBase={
          <input
            checked={false}
            onChange={[Function]}
            singleCheckboxRadio={true}
            type="radio"
          />
        }
        className="checkbox-radio-label radio-wrapper"
        inline={false}
        inputType="radio"
        labelAlign="right"
        onChange={[Function]}
        wrapperClassName="col-xs-offset-2 col-xs-10"
      >
        <label
          className="checkbox-radio-label radio-wrapper"
        >
          <input
            checked={false}
            onChange={[Function]}
            singleCheckboxRadio={true}
            type="radio"
          />
        </label>
      </Label>
    </div>
  </LayoutHelper>
</Radio>
`;

exports[`Radio with default props should render corrently 1`] = `
<Radio
  bsClass="radio"
  inline={false}
  inputType="radio"
  labelAlign="right"
  onChange={[Function]}
  singleCheckboxRadio={true}
  standalone={true}
  wrapperClassName="col-xs-offset-2 col-xs-10"
>
  <LayoutHelper
    FieldBase={
      <input
        onChange={[Function]}
        singleCheckboxRadio={true}
        type="radio"
      />
    }
    bsClass="radio"
    checked={false}
    inline={false}
    inputType="radio"
    labelAlign="right"
    onChange={[Function]}
    singleCheckboxRadio={true}
    standalone={true}
    wrapperClassName="col-xs-offset-2 col-xs-10"
  >
    <div
      className="radio"
    >
      <Label
        FieldBase={
          <input
            onChange={[Function]}
            singleCheckboxRadio={true}
            type="radio"
          />
        }
        className="checkbox-radio-label radio-wrapper"
        inline={false}
        inputType="radio"
        labelAlign="right"
        onChange={[Function]}
        wrapperClassName="col-xs-offset-2 col-xs-10"
      >
        <label
          className="checkbox-radio-label radio-wrapper"
        >
          <input
            onChange={[Function]}
            singleCheckboxRadio={true}
            type="radio"
          />
        </label>
      </Label>
    </div>
  </LayoutHelper>
</Radio>
`;
